librz/analysis/var_global.c

```diff
@@ -0,0 +1,370 @@
+	return ht_pp_update_key(analysis->ht_global_var, old_name, newname);
+}
+
+/**
+ * \brief Set the type of the global variable
+ * 
+ * \param glob Global variable instance
+ * \param type The type to set. RzType*
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_set_type(RzAnalysisVarGlobal *glob, RzType *type) {
+	rz_return_if_fail(glob && type);
+	glob->type = type;
+}
+
+/**
+ * \brief Add a constaint to global variable
+ * 
+ * \param glob Global variable instance
+ * \param constraint RzTypeConstraint
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_add_constraint(RzAnalysisVarGlobal *glob, RzTypeConstraint *constraint) {
+	rz_return_if_fail(glob && constraint);
+	rz_vector_push(&glob->constraints, constraint);
+}
+
+/**
+ * \brief Get the pritable string of global variable constraints
+ * 
+ * \param glob Global variable instance
+ * \return char *
+ */
+RZ_API char *rz_analysis_var_global_get_constraints_readable(RzAnalysisVarGlobal *glob) {
+	size_t n = glob->constraints.len;
+	if (!n) {
+		return NULL;
+	}
+	bool low = false, high = false;
+	RzStrBuf sb;
+	rz_strbuf_init(&sb);
+	size_t i;
+	for (i = 0; i < n; i += 1) {
+		RzTypeConstraint *constr = rz_vector_index_ptr(&glob->constraints, i);
+		switch (constr->cond) {
+		case RZ_TYPE_COND_LE:
+			if (high) {
+				rz_strbuf_append(&sb, " && ");
+			}
+			rz_strbuf_appendf(&sb, "<= 0x%" PFMT64x "", constr->val);
+			low = true;
+			break;
+		case RZ_TYPE_COND_LT:
+			if (high) {
+				rz_strbuf_append(&sb, " && ");
+			}
+			rz_strbuf_appendf(&sb, "< 0x%" PFMT64x "", constr->val);
+			low = true;
+			break;
+		case RZ_TYPE_COND_GE:
+			rz_strbuf_appendf(&sb, ">= 0x%" PFMT64x "", constr->val);
+			high = true;
+			break;
+		case RZ_TYPE_COND_GT:
+			rz_strbuf_appendf(&sb, "> 0x%" PFMT64x "", constr->val);
+			high = true;
+			break;
+		default:
+			break;
+		}
+		if (low && high && i != n - 1) {
+			rz_strbuf_append(&sb, " || ");
+			low = false;
+			high = false;
+		}
+	}
+	return rz_strbuf_drain_nofree(&sb);
+}
+
+/**
+ * \brief Print out the global variables
+ * 
+ * \param analysis RzAnalysis
+ * \param state RzCmdStateOutput *
+ * \param name Which one to print
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_list_show(RzAnalysis *analysis, RzCmdStateOutput *state, RZ_NULLABLE const char *name) {
+	rz_return_if_fail(analysis && state);
+	RzList *global_vars = NULL;
+	RzAnalysisVarGlobal *glob = NULL;
+	if (name) {
+		global_vars = rz_list_new();
+		if (!global_vars) {
+			return;
+		}
+		glob = rz_analysis_var_global_get_byname(analysis, name);
+		if (!glob) {
+			return;
+		}
+		rz_list_append(global_vars, glob);
+	} else {
+		global_vars = rz_analysis_var_global_get_all(analysis);
+	}
+
+	RzListIter *it = NULL;
+	char *var_type = NULL;
+	bool json = state->mode == RZ_OUTPUT_MODE_JSON;
+	PJ *pj = json ? state->d.pj : NULL;
+	// to use rz_cmd_state_output_array_start we need to set RzCore as the dependency of RzAnalysis, which is impossible
```

Comment:
To solve this, we could just move this particular function to `librz/core/canalysis.c` instead. We tend to have printing functions there.

Reply:
This could be done in another PR.

Reply:
next PR